Do not render spree/shared/social if no more authentication methods available for user

westonganger opened the pull request.

westonganger committed:
Do not render spree/shared/social if no more authentication methods available for user.

add tests, fix spree/shared/_social if condition.

fix incorrect use of @spree_user.

jarednorman commented on app/views/spree/shared/_social.html.erb:
```diff
@@ -3,7 +3,7 @@
   <% end %>
 
-  <% Spree::AuthenticationMethod.available_for(@spree_user).each do |method| %>
+  <% if Spree::AuthenticationMethod.available_for(@spree_user).each do |method| %>
```
`.each` returns the original collection, so this condition will never be falsy.

jarednorman commented:
Perhaps I'm reading this incorrectly, but I don't see how this PR changes anything.

tvdeyen commented:
Thanks for your contribution, but Jared is right. An empty collection will still be true.

You probably want to put this in a `.present?` block first. 

Also, please add tests.

westonganger commented:
It is still showing the Facebook icon even though I have already used that method of authentication and `exists?` returns false. See the attached image for reference. I will add the test shortly.
![screenshot-127 0 0 1 3000 2016-04-11 08-06-01](https://cloud.githubusercontent.com/assets/3414795/14431845/64f47826-ffbc-11e5-95b6-d5f3381a3cc0.png)

jarednorman commented on app/views/spree/shared/_social.html.erb:
```diff
@@ -2,10 +2,12 @@
-          spree.spree_user_omniauth_authorize_url(provider: method.provider),
-          title: Spree.t(:sign_in_with, provider: method.provider)) if method.active %>
+  
+  <% if Spree::AuthenticationMethod.available_for(spree_current_user).exists? %>
```
Adding an outer conditional doesn't change the behaviour as far as I can tell. If there aren't any, the loop never runs, and since this partial isn't rendered if there aren't any, it's extra not needed.

jarednorman commented on app/views/spree/users/_social.html.erb:
```diff
@@ -1,28 +1,30 @@
+      <% end %>
     <% end %>
+    
+    <% if Spree::AuthenticationMethod.available_for(spree_current_user).exists? %>
```
I'm fairly sure `.exists?` will run an additional query that won't hit the query cache when `.available_for` is used inside the partial. `.present?` or something would be preferable.

jarednorman commented:
Thanks for the contribution. Left a couple of small comments, but beyond those, this looks good to merge to me.

westonganger commented:
I cleaned this up and squashed it into one commit. I think the main problem why it was showing the icons was because it was calling `@spree_user` which should have been `spree_current_user`

jarednorman commented:
Looking good. The specs that are failing are the same as the ones on master (likely due to admin changes which I will look into shortly) so I think this is good to merge.

jhawthorn commented:
Thank you
